Lib/test/test_typing.py

```
@@ -9097,6 +9123,71 @@ class AllTheThings(TypedDict):
+        self.assertEqual(Child.__mutable_keys__, frozenset({'a'}))
+        self.assertEqual(Child.__annotations__, {"a": str})
+        self.assertIs(Child.__extra_items__, int)
+        self.assertIsNone(Child.__closed__)
+
+        class GrandChild(Child, closed=True):
+            a: float
+
+        self.assertEqual(GrandChild.__required_keys__, frozenset({'a'}))
+        self.assertEqual(GrandChild.__optional_keys__, frozenset({}))
+        self.assertEqual(GrandChild.__readonly_keys__, frozenset({}))
+        self.assertEqual(GrandChild.__mutable_keys__, frozenset({'a'}))
+        self.assertEqual(GrandChild.__annotations__, {"a": float})
+        self.assertIs(GrandChild.__extra_items__, NoExtraItems)
+        self.assertIs(GrandChild.__closed__, True)
+
+        class GrandGrandChild(GrandChild):
+            ...
+        self.assertEqual(GrandGrandChild.__required_keys__, frozenset({'a'}))
+        self.assertEqual(GrandGrandChild.__optional_keys__, frozenset({}))
+        self.assertEqual(GrandGrandChild.__readonly_keys__, frozenset({}))
+        self.assertEqual(GrandGrandChild.__mutable_keys__, frozenset({'a'}))
+        self.assertEqual(GrandGrandChild.__annotations__, {"a": float})
+        self.assertIs(GrandGrandChild.__extra_items__, NoExtraItems)
+        self.assertIsNone(GrandGrandChild.__closed__)
+
+    def test_implicit_extra_items(self):
+        class Base(TypedDict):
+            a: int
+
+        self.assertIs(Base.__extra_items__, NoExtraItems)
+        self.assertIsNone(Base.__closed__)
+
+        class ChildA(Base, closed=True):
+            ...
+
+        self.assertEqual(ChildA.__extra_items__, NoExtraItems)
+        self.assertIs(ChildA.__closed__, True)
+
+    def test_cannot_combine_closed_and_extra_items(self):
+        with self.assertRaisesRegex(
+            TypeError,
+            "Cannot combine closed=True and extra_items"
+        ):
+            class TD(TypedDict, closed=True, extra_items=range):
```

Is `range` the builtin here?
That's a bit strange.

---

I agree it looks a bit odd as an arbitrary `extra_items` value, but it's valid, since it just means extra keys must have values assignable to [range](https://docs.python.org/3/library/stdtypes.html#sequence-types-list-tuple-range). For context, this test is copied from `typing_extensions`, where it was [used](https://github.com/python/typing_extensions/blob/abaaafd98c1cc7e5baf098ec287a3d22cb339670/src/test_typing_extensions.py#L5179) to cover the same case. We could edit it for readability, but I kind of like it as a reminder for maintainers that `extra_items` accepts any type, not just the obvious ones.

---

That could be valid for a regular test that validates stored values.

this however is a negative test, isn’t it?

---

It's still valid/correct as a negative test. By "valid" do you mean "correct" or "readable?" Passing range into extra_items is valid, and passing closed=True at the same time should raise a runtime error.

---

Sorry I was being terse.

I'm trying to say that this test:

```py
        class Child(Base, extra_items=int):
            a: str

        self.assertEqual(Child.__required_keys__, frozenset({'a'}))
        self.assertEqual(Child.__optional_keys__, frozenset({}))
        self.assertEqual(Child.__readonly_keys__, frozenset({}))
        self.assertEqual(Child.__mutable_keys__, frozenset({'a'}))
        self.assertEqual(Child.__annotations__, {"a": str})
        self.assertIs(Child.__extra_items__, int)
        self.assertIsNone(Child.__closed__)
```

Could have an `extra_items=range` counterpart.

That would make a solid test, both understandable and useful.

Meanwhile, the negative test, `class TD(TypedDict, closed=True, extra_items=range): --> error` would be better served with a simpler, more straightforward `extra_items=int` argument.

---

P.S. my comment overall is minor, please don't let me stop your work!

---

Thanks for clarifying. I thought about it some more.

I don't think introducing an `extra_items=range` counterpart would actually widen the test coverage, since it wouldn't be exercising any new behavior. PEP 728 splits responsibilities between runtime and type checker behavior. While `extra_items` is only supposed to accept a valid [type expression](https://typing.python.org/en/latest/spec/glossary.html#term-type-expression), validating that is the **type checker's** job (e.g. MyPy's [valid-type](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-validity-of-types-valid-type) error), not the runtime's. The runtime just stores whatever is passed in.

So the real subject under test is simply:

> `Child.__extra_items__` must correctly store the value passed into `extra_items`.

We don't need multiple values to prove that behavior.

And although `range` is a less obvious type, I think it makes sense in the negative test. That test is specifically asserting the error message “Cannot combine `closed=True` and `extra_items`”. Using `range` highlights that the failure comes from the combination, not from `range` itself being an invalid value of `extra_items`.

---

I tend to use `range` sometimes as it's a builtin type that isn't generic (like `list`) and doesn't participate in promotion weirdness (like `float` and historically `bytes`), so it's a good basic type to test with.

Plus, people who forget that `range` is a type get to learn something :)